Initial implementation for SAP connector (#1)
wso2-ballerina/module-sap: README.md

# module-sap

## Package overview

> Ballerina SAP Client Endpoint is used to connect Ballerina with SAP. With this SAO Client Endpoint, Ballerina can act as SAP Consumers and SAP Producers.

Reviewer: ```suggestion
Ballerina SAP Client Endpoint is used to connect Ballerina with SAP. With this SAP Client Endpoint, Ballerina can act as SAP Consumers and SAP Producers.
```

> Ballerina SAP Client Endpoint is used to connect Ballerina with SAP. With this SAO Client Endpoint, Ballerina can act as SAP Consumers and SAP Producers.

Reviewer 2: I believe we should reword this since we have both a client and a listener in this module.

Basically we have to say that the Ballerina SAP Client acts as a SAP producer while the Ballerina SAP Listener acts as a SAP Consumer.

IMO, we don't have to use the "endpoint" word here.

    
    <table class="tg">
      <tr>
        <td class="tg-yw4l">Linux 32-bit</td>
        <td class="tg-yw4l">Copy the Linux native SAP jcolibrary libsapjco3.so to <JDK_HOME>/jre/lib/i386/server.</td>
      </tr>
      <tr>
        <td class="tg-yw4l">Linux 64-bit</td>
        <td class="tg-yw4l">Copy the Linux native SAP jcolibrary libsapjco3.so to <JDK_HOME>/jre/lib/amd64.</td>
      </tr>
      <tr>
        <td class="tg-yw4l">Windows</td>
        <td class="tg-yw4l">Copy the Windows native SAP jcolibrary sapjco3.dll to <WINDOWS_HOME>/system32.
        </td>
      </tr>
    </table>
    
 * Extract wso2-sap-<version>.zip and  Run the install.sh script to install the module.
  
   You can uninstall the module by running uninstall.sh.

## Building From the Source

If you want to build Ballerina sap endpoint from the source code:

1. Get a clone or download the source from this repository:
    [https://github.com/wso2-ballerina/module-sap](https://github.com/wso2-ballerina/module-sap)
2. Create lib folder into the module-sap 
3. Download the sapidoc3.jar and sapjco3.jar middleware libraries from the SAP support portal and copy those 
   libraries to the module-sap/lib

> 4. Run the following Maven command from the ballerina directory:

Reviewer 3: ```suggestion
4. Navigate to the <BALLERINA_HOME> directory and execute the following Maven command:
```

    mvn clean install

> 5. Extract the distribution created at `/component/target/wso2-sap-<version>.zip`. Run the install.{sh/bat} script to install the module.

Reviewer 3: ```suggestion
5. Extract the distribution created at `/component/target/wso2-sap-<version>.zip`.
6. Execute the install.{sh/bat} script to install the module.
```

> You can uninstall the module by running uninstall.{sh/bat}.    

Reviewer 3: ```suggestion
You can uninstall the module by running either of the uninstall.{sh/bat} scripts.     
```

      
## Samples

### SAP Producer

> Following example demonstrates a way to publish an idoc to sap.

Reviewer 3: ```suggestion
The following example demonstrates how to publish an IDoc to SAP.
```

```ballerina
import wso2/sap;
import ballerina/io;

sap:ProducerConfig producerConfigs = {
    destinationName:"<The SAP gateway name>",
    sapClient:"<SAP client, for example, 001>",
    userName:"<The user logon>",
    password:"<The logon password>",
    asHost:"<The R/3 application server>",
    sysnr:"<SAP system number, for example, 01>",
    language:"<The logon language>"
};

sap:Producer sapProducer = new(producerConfigs);

string qname = "_-DSD_-ROUTEACCOUNT_CORDER002";
int idocVersion = The version of IDoc;
xml idoc = xml `<IDoc Data>`; 
NOTE: Here, If Any tag in the IDoc start with "_-", then you have to intialise that tag and use it in the IDoc. 
      eg: <{{qname}}>...</{{qname}}>  

public function main() {

>     var result = sapProducer-> sendIdocMessage(idocVersion, idoc);

Reviewer 4: ```suggestion
    var result = sapProducer->sendIdocMessage(idocVersion, idoc);
```

    if (result is error) {
        io:println("Error: ", result.reason());
    } else {
        io:println("Response: ", result);
    }
}
````

> Following example demonstrates a way to publish a bapi to sap.

Reviewer 2: IINM, the convention is to use upper case chars with BAPI and SAP. And IDoc with just I and D capital. Shall we fix all documentation/comments?

> Following example demonstrates a way to publish a bapi to sap.

Reviewer 3: ```suggestion
The following example demonstrates how to publish a BAPI to SAP.

```

```ballerina
xml baPiData = xml `<BAPI Data>`;
public function main() {

>     var result = sapProducer-> sendBapiMessage(bapiData , false, false);

Reviewer 4: ```suggestion
    var result = sapProducer->sendBapiMessage(bapiData , false, false);
```

    if (result is error) {
        io:println("Error: ", result.reason());
    } else {
        io:println("Response: ", result);
    }
}
```

### SAP Consumer

> ```ballerina

Reviewer: Shall we reformat the code sample properly, as below?

```ballerina
import wso2/sap;
import ballerina/io;

listener sap:Listener consumerEP = new({
        transportName: "<The protocol name[idoc/bapi]>",
        serverName: "<Name of the server configuration>",
        gwHost: "<Gateway host on which the server should be registered>",
        progId: "<The program ID with which the registration is done>",
        repositoryDestination: "<Name of the .dest file>",
        gwServ: "<Gateway service>",
        unicode: "<Determines whether or not you connect in unicodemode>" }, {
        sapClient: "<SAP client, for example, 001>",
        userName: "<The user logon>",
        password: "<The logon password>",
        asHost: "<The R/3 application server>",
        sysnr: "<SAP system number, for example, 01>",
        language: "<The logon language>"
    }
);

service SapConsumerTest on consumerEP {
    // The `resource` registered to receive server messages
    resource function onMessage(string idocList) {
        io:println("The message received from SAP instance : " + idocList);
    }
    // The `resource` registered to receive server error messages
    resource function onError(error err) {
        io:println("Error from Connector: " + err.reason());
    }
}
```
Please reformat the other code samples as well.

import wso2/sap;
import ballerina/io;

listener sap:Listener consumerEP = new ({
    transportName:"<The protocol name[idoc/bapi]>",
    serverName:"<Name of the server configuration>",
    gwHost:"<Gateway host on which the server should be registered>",
    progId:"<The program ID with which the registration is done>",
    repositoryDestination:"<Name of the .dest file>",
    gwServ:"<Gateway service>",
    unicode:"<Determines whether or not you connect in unicodemode>"}, {
    sapClient:"<SAP client, for example, 001>",
    userName:"<The user logon>",
    password:"<The logon password>",
    asHost:"<The R/3 application server>",
    sysnr:"<SAP system number, for example, 01>",
    language:"<The logon language>"
    }

>     );

Reviewer 2: Formatting could improve. Let's indent appropriately.

service SapConsumerTest on consumerEP {
    // The `resource` registered to receive server messages

>     resource function onMessage(string idocList) {

Reviewer 2: Could you share a sample `idocList` that could be received? IIRC, we discussed for this to be an XML right? Is it a different structure?
